### ISSUE TITLE
get string width

### ISSUE THREAD

**Issue by Yardie-** [k-omura/truetype_Arduino#6]
Good idea I will follow your lead
Two new suggestions.
If x y co-ordinates are int16_t instead of uint16_t they allow off screen co-ordinates ie < 0
The way I handle this is just do a simple bounds check in the addPixel method. Then just return if the pixel is out of bounds.
this allows strings to easily be drawn right to the edge of the screen even from the start.
It also allows rotation in any of the 4 directions 
with a simple (this is from my code but you will get the idea)

```
int16_t temp = page_x;
  int16_t canvas_x, canvas_y;
  if( canvas->rotation == Rotate_90 )
  {
    canvas_x = canvas->width_in_pixels -1 - page_y ;
    canvas_y =  temp;
  }
  else if( canvas->rotation == Rotate_180 )
  {
    canvas_x = canvas->width_in_pixels -1 - page_x;
    canvas_y = canvas->height_in_pixels -1 - page_y;
  }
  else if( canvas->rotation == Rotate_270 )
  {
    canvas_x = page_y ;
    canvas_y = canvas->height_in_pixels -1 - temp ;
  }
  else
  {
    canvas_y = page_y;
    canvas_x = page_x;
  }
  // now all is rotated test out of bounds
  if (canvas_x < 0 || (uint16_t)canvas_x >= canvas->width_in_pixels || canvas_y < 0 || (uint16_t) canvas_y >= canvas->height_in_pixels)
    return false;  
// now we are free to set the pixel 

```

However the main suggestion here is to allow simple page centring of the string.
It is a simple cut and paste of your code. With a couple of minor changes.
I am sure you can come up with a cleaner solution.

thus we can call 
x = buffer_width /  2  - getStringWidth(...)  /  2;
then 
string( x...
will centre the string on the page.

```
uint16_t truetypeClass::getStringWidth(uint16_t _x, uint16_t _y, const wchar_t _character[]){
  uint8_t c = 0;
  uint16_t prev_code = NULL;

  while (_character[c] != '\0') {
    //space (half-width, full-width)
    if((_character[c] == ' ') || (_character[c] == '　')){
      prev_code = NULL;
      _x += this->characterSize / 4;
      c++;
      continue;
    }
    //Serial.printf("%c\n", _character[c]);

    uint16_t code = this->codeToGlyphId(_character[c]);
    this->readGlyph(code);

    _x += this->characterSpace;
    if(prev_code != NULL && this->kerningOn){
      int16_t kern = this->getKerning(prev_code, code); //space between charctor
      _x += (kern * (int16_t)this->characterSize) / (this->yMax - this->yMin);
    }
    prev_code = code;


    ttHMetric_t hMetric = getHMetric(code);
    uint16_t width = this->characterSize * (glyph.xMax - glyph.xMin) / (this->yMax - this->yMin);

    //Line breaks when reaching the edge of the display
    if((hMetric.leftSideBearing + width + _x) > this->end_x){
      _x = this->start_x;
      _y += this->characterSize;
      if(_y > this->end_y){
        break;
      }
    }

    this->freePointsAll();
    this->freeGlyph();

    _x += (hMetric.advanceWidth) ? (hMetric.advanceWidth) : (width);
    c++;
  }
  return x;
}

```

Sorry for all my edits

**Comment by k-omura** [comment]
Many suggestions are appreciated!
It has led to an essential library improvement!

About the first suggestion:
I think it is absolutely necessary to check the negative coordinates.
(Same for other error checks)
It will be little by little, but I would like to update it.

I think rotation is also convenient. Supporting rotation may lead to increased support for the framebuffer format.

About the second suggestion:
I thought it would be convenient if I could align the center and right!
As you say, you can easily get the string width except for the part that writes to the framebuffer from the current string function, so it seems easy to implement.
(Although it takes a little longer than left alignment)

My holiday will be over soon, so it may take some time, but I want to move on 😇

**Comment by Yardie-** [comment]
A lot of the code I have worked with around epapers use unsigned values then throw out the hole thing when it's out of bounds.
I have found that doing the bounds checking in the actual pixel set section removes the need to worry about it elsewhere.
If a string is too long it just doesn't get rendered. 
Yes it's longer but if the user wants to speed it up they can just send smaller strings or make other adjustments.
Speed is more of an issue with LCD's but with epapers especially if you use both cores it's not such and issue.

The rotation is fun especially with a frame buffer or two.
Also having a clean straight forward library with a clear and straight forward API is a bonus.
I have made lots of suggestions but they are really only to get a functional flexible core.
Often people try to do too much and the examples get way to complicated. Then you can't see the wood for the trees. I totally agree with your idea of sticking to the frame buffer. Actually it may be better to just do my DES code so it hooks in to your API but is completely separate.
The way you have handled the SPIFFS SD file source issue. It becomes a non issue then.
Give me a file handle, a file, a uint8_t array, and a string and I'll give you an image. Very NICE

I will hack out a two core example when I get my DES screens and get that code working. The epaper does it's thing on one core while the other core is collecting data or controlling time critical functions. Once you understand them they are really easy to implement. Simplest version of multi threading I have ever worked with.  
Once people understand that I think epaper will become more popular. The documentation is still so vague it makes it hard for people to get started.

I still don't quite understand git. I will work it out so I can make clean examples that allow people to use your code without driving you crazy with unrelated questions.

I understand you have limited time. I am very grateful and will help where I can.

**Comment by k-omura** [comment]
I added and committed the main contents of this issue, error checking of addPixel and getting the width of the string.  (Let's handle the rotation with # 8)
The width of the string is almost as you thought, but I changed it to read only the data necessary for this purpose from the ttf file.
I think it's okay to close this issue here. Close if there are no additional comments.
Thank you for the idea.

**Comment by Yardie-** [comment]
funny we are working on the same thing but as always it's your call.
Have a look at what I did with addPixel.
It now handles textBoundaries nicely 
I've solved the problems with rotation as well.
The way I am looking at stringwidth is that it adds the position it is upto on each virtual draw. 
Thus you have the possition in the string of the last character drawn 
move down a line and start again.

This could if you like have the " " character add position of last word thus it would be easier to break on words and move down a line as you know which string position the last char is.
I stuck to wchar strings because that is effectively what the code is dealing with

with the last word idea you could copy the string up to that point and 
rerun get width and you would have centered word broken strings for each line.
I also made it return on a carriage return or line feed char.
*nix or mac windows will return twice but that could be handled differently
by the user.

**Comment by k-omura** [comment]
Certainly, it makes sense to start a new line or measure the width of a string with CR or LF.
In essence, these implementations are probably very easy, and I plan to implement them soon, although there are differences depending on the environment.

I understand that it is also important in space-based languages ​​to pre-break words after spaces as needed.
I also want to take care of this in the future.
This is actually not very important in Japanese. Because it's not a space-based language. However, space is often used for clarity.
Therefore, it may be difficult to deal with them centrally.
Pre-line breaks in non-space-based languages ​​result in a large number of useless line breaks.
I haven't thought about any concrete solution yet, but I want to think about it.

I've confirmed your code and comments, but I haven't been able to catch up completely. Sometimes I may commit code that is different from your idea, but I'm trying to absorb your thoughts and improve it in a way that doesn't affect the user API.

Best regards in the future

**Comment by Yardie-** [comment]
That is why I went ahead with my branch.
I am after code I can work with and that makes sense to me. 
If the code doesn't match but is close enough and it works then both versions can run in parallel and we can both benefit from each others coding and cultural understanding. 
I added a few defines (I wasn't sure if they would work as they do in C but they seem to) as I would like to use this as easily in esp-idf or on the pi (which don't have String which I think is the only incompatibility).

Also I had an idea I haven't tested yet.
Does the code still break if the framebuffer is global but the class is contained.

Anyway the object is to have fun and get something out of this in the end. 
In that regard I am succeeding and I hope you are too.
Talk soon 
Thanks K cpp is not that bad after all. :)
(Sorry I don't know your name but K like in "Men in black" works)

Greg
